All chat logs beside Q8 and Q7.4. can be found here (including a summary):
https://chatgpt.com/share/a1b8a72e-bdd7-45d5-8e06-8de1e5790fc8

In [3]:
#1.

import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

1. 
This shows that the imported data has missing values because the function "isna" checks if each value is missing or availible and the code prints the sum of the missing values. Since there are number greater than zero, we know that there are missing values in this dataset.

In [4]:
#2.1
import pandas as pd

# Load the dataset from the URL
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_df = pd.read_csv(url)

# Get the shape of the dataset (rows, columns)
villagers_shape = villagers_df.shape

print(f'This dataset has {villagers_shape[0]} rows and {villagers_shape[1]} columns.')

This dataset has 391 rows and 11 columns.


2.2
Observation refers to a single row, in my dataset observation refers to a villager from the game "Animal Crossing".
Variable refers to a single column, in my dataset variable refers to a charactistic for the villagers in the game "Animal Crossing".
The link to the chat logs with chatgpt with a summary at the end of it: https://chatgpt.com/share/a1b8a72e-bdd7-45d5-8e06-8de1e5790fc8 

In [1]:
#3.
import pandas as pd

# Load the dataset from the URL
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_df = pd.read_csv(url)

# Display descriptive statistics for numerical columns
numeric_summary = villagers_df.describe()

# Display counts of unique values in each categorical column
categorical_summary = {col: villagers_df[col].value_counts() for col in villagers_df.columns if villagers_df[col].dtype == 'object'}

# Print the summaries
print("Descriptive Statistics for Numeric Columns:\n", numeric_summary)
print("\nCounts of Unique Values in Categorical Columns:")
for col, counts in categorical_summary.items():
    print(f"\nColumn: {col}\n{counts}")

Descriptive Statistics for Numeric Columns:
             row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

Counts of Unique Values in Categorical Columns:

Column: id
id
admiral    1
mott       1
paula      1
patty      1
pate       1
          ..
eloise     1
elmer      1
ellie      1
elise      1
zucker     1
Name: count, Length: 390, dtype: int64

Column: name
name
Admiral    1
Muffy      1
Paula      1
Patty      1
Pate       1
          ..
Elvis      1
Eloise     1
Elmer      1
Ellie      1
Zucker     1
Name: count, Length: 391, dtype: int64

Column: gender
gender
male      204
female    187
Name: count, dtype: int64

Column: species
species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf

In [4]:
#4.
import pandas as pd

# Load the Titanic dataset
titanic= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic_df = pd.read_csv(titanic)

# Get the shape of the dataset (total number of rows and columns)
shape = titanic_df.shape

# Get the descriptive statistics for numerical columns
describe = titanic_df.describe()

# Print the results
print("Shape of the dataset (rows, columns):", shape)
print("\nDescriptive statistics for numerical columns:\n", describe)

Shape of the dataset (rows, columns): (891, 15)

Descriptive statistics for numerical columns:
          survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200


Using this code we can see the the column age only has 714 data points, however, we know by seeing the shape that there should be 891. Therefore the column age is missing 177 data points.

5.
Attribute: Represents a characteristic or the state of an object or thing.
Method: Is a function which shows what actions the object or thing can do.

6.(copied from chatgpt)

count: The number of non-null (non-missing) values in the column.
mean: The average value of the column.
std: The standard deviation, which measures the spread of the data around the mean.
min: The minimum value in the column.
25% (1st quartile): The value at the 25th percentile, meaning 25% of the data lies below this value.
50% (median): The median value, which is the middle value when the data is sorted.
75% (3rd quartile): The value at the 75th percentile, meaning 75% of the data lies below this value.
max: The maximum value in the column.

7.1. A case that df.dropna() would be preferreed instead of del df['col'] is when we need to remove rows with missing data in muliple columns. For example, if a row from a survey taker did not answer 1 or more questions, the data shown could be influenced by the people who did answer all the questions, therefore it would be better to use only people who answered all the question to have equal amounts of answers for each catergory. 

7.2. A case that del df['col'] would be preffered instead of df.dropna() would be when you have a column that is unneccesary is is often blank. For example, if a question like "What's you favourite animal?" is often left blank, that column may need to be removed as there is no need to have it as it may be less important compared to other questions.

7.3. When both del df['col'] and df.dropna() are going to be used, would be better to use del df['col'] before df.dropna() because if a column or multiple columns are commonly missing values, then we would want to delete them because if we dropped the rows first we would lose a lot more rows than one or a few columns.

7.4. The approach I use in the following code uses only df['col'] since in my url there are only 4 columns that have missing data "age", "embarked", "deck", and embarked_town".

The chat with Chatgpt for this question(7.4.):
https://chatgpt.com/share/c1331f78-7e0a-491a-872a-e378cd83f312

In [5]:
import pandas as pd

# Load the Titanic dataset from the URL
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Check for missing data before cleaning
missing_data_before = df.isnull().sum()
print("Missing data before cleaning:")
print(missing_data_before)

# Remove columns with any missing values
for col in df.columns:
    if df[col].isnull().any():
        del df[col]

# Check for missing data after cleaning
missing_data_after = df.isnull().sum()
print("\nMissing data after cleaning:")
print(missing_data_after)


Missing data before cleaning:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

Missing data after cleaning:
survived      0
pclass        0
sex           0
sibsp         0
parch         0
fare          0
class         0
who           0
adult_male    0
alive         0
alone         0
dtype: int64


8.1. The chatgpt chat log for all of Q8. can be found here(including a summary at the end): 
https://chatgpt.com/share/ce68c70c-fff8-464f-87de-b9e9d4f7c542

I asked python to explain to me what df.groupby("col1")["col2"].describe() does and an example I have used is taking "pclass" as "col1" and "survived" as "col2" from the file "titanic" to compare these 2 columns which shows the closer to 1st pclass the better your survivle rate would be on the titanic.

In [7]:
#This is the code I changed for my example
import pandas as pd

# Load the Titanic dataset from the URL
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'pclass' and describe the 'survived' column
grouped_stats = df.groupby("pclass")["survived"].describe()
print(grouped_stats)

        count      mean       std  min  25%  50%  75%  max
pclass                                                    
1       216.0  0.629630  0.484026  0.0  0.0  1.0  1.0  1.0
2       184.0  0.472826  0.500623  0.0  0.0  0.0  1.0  1.0
3       491.0  0.242363  0.428949  0.0  0.0  0.0  0.0  1.0


8.2. The count in df.describe() shows the count for all the non-null values in that column, however df.groupby("col1")["col2"].describe() counts how many non-null values from "col2" defined by "col1" there are.

In [1]:
#8.3.A. 

# Load the dataset from the URL
url = "titanic.csv"
df = pd.read_csv(url)

# Get the shape of the dataset (rows, columns)
titanic_shape = df.shape

print(f'This dataset has {titanic_shape[0]} rows and {titanic_shape[1]} columns.')

NameError: name 'pd' is not defined

After the first time entering the error into chatgpt, clearly shows its better than google.

In [19]:
#8.3.B. 

import pandas as pd

# Load the Titanic dataset
url = "titanics.csv"
df = pd.read_csv(url)

# Get the shape of the dataset (rows, columns)
titanic_shape = df.shape

print(f'This dataset has {titanic_shape[0]} rows and {titanic_shape[1]} columns.')

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

In the chatgpt logs it shows that it says that mispelling the file is a possibility, which would then make the user(me) check if it was spelled correctly, which is better than google.

In [11]:
#8.3.C. 

import pandas as pd

# Load the Titanic dataset
url = "titanic.csv"
df = pd.read_csv(url)

# Get the shape of the dataset (rows, columns)
titanic_shape = df.shape
DF.groupby("pclass")["survived"].describe()




NameError: name 'DF' is not defined

Once again, chatgpt is telling me that the variable should be df which is a lot better than any google searches.

In [3]:
#8.3.D. 

import pandas as pd

# Load the Titanic dataset
url = "titanics.csv"
df = pd.read_csv(url

# Get the shape of the dataset (rows, columns)
titanic_shape = df.shape

print(f'This dataset has {titanic_shape[0]} rows and {titanic_shape[1]} columns.')

SyntaxError: '(' was never closed (2551714266.py, line 7)

Chat gpt told me exactly what to do to fix this, but google also said how to say it in my first search. However, chatgpt gave me an example of code to fix it, which google did not.

In [12]:
#8.3.E.

import pandas as pd

# Load the Titanic dataset
url = "titanic.csv"
df = pd.read_csv(url)

# Get the shape of the dataset (rows, columns)
titanic_shape = df.shape
df.groupby('pclass')["survived"].describle()

AttributeError: 'SeriesGroupBy' object has no attribute 'describle'

Chat gpt tells me that the correct method is .describe() instead of .describle(). Google makes me look through websites to evnetually fine that the method is .describe() not .describle().

In [15]:
#8.3.F.

import pandas as pd

# Load the Titanic dataset
url = "titanic.csv"
df = pd.read_csv(url)

# Get the shape of the dataset (rows, columns)
titanic_shape = df.shape
df.groupby("Sex")["age"].describe()

KeyError: 'Sex'

Chatgpt once again tells me that i probably have misspelled this while google just tells me I need to put a column name there instead of "Sex".

In [16]:
#8.3.G.

import pandas as pd

# Load the Titanic dataset
url = "titanic.csv"
df = pd.read_csv(url)

# Get the shape of the dataset (rows, columns)
titanic_shape = df.shape
df.groupby(sex)["age"].describe()

NameError: name 'sex' is not defined

Chatgpt then tells me to make it "Sex" however this wouldn't work but if I put that into chatgpt it would result as the same answer for 8.3.F. which is still better than google.

9. Somewhat